core/src/main/scala/kafka/log/LogValidator.scala

```diff
@@ -78,6 +78,12 @@ private[kafka] object LogValidator extends Logging {
       if (batch.isControlBatch)
         throw new InvalidRecordException("Clients are not allowed to write control records")
     }
+
+    if (batch.isTransactional && toMagic < RecordBatch.MAGIC_VALUE_V2)
+      throw new UnsupportedForMessageFormatException(s"Transactional records cannot be used with magic version $toMagic")
```

Comment:
The `UnsupportedForMessageFormatException` should get caught by the request handler and returned as an error code to the user, right? That is why I updated the response code for the `ProduceRequest`.